sdk/eventhub/azure-eventhub/azure/eventhub/aio/_transport/_pyamqp_transport_async.py

```diff
@@ -161,6 +161,7 @@ async def _callback_task(consumer, batch, max_batch_size, max_wait_time):
             events = [EventData._from_message(message) for message in messages] # pylint: disable=protected-access
             now_time = time.time()
             if len(events) > 0:
+                consumer._last_received_event = events[-1] # pylint: disable=protected-access
```

Comment:
Could link me to the equivalent mechanism in the sync/uamqp transports?

Reply:
https://github.com/l0lawrence/azure-sdk-for-python/blob/9e76c9b490b09677140cdad55b03229532a0260b/sdk/eventhub/azure-eventhub/azure/eventhub/_consumer.py#L188

Reply:
https://github.com/l0lawrence/azure-sdk-for-python/blob/9e76c9b490b09677140cdad55b03229532a0260b/sdk/eventhub/azure-eventhub/azure/eventhub/aio/_consumer_async.py#L181 -- Uamqp Async

Reply:
hmm it seems that for every other implementation this was handled at the consumer level. Would it make sense that pyamqp be fixed so that the consumer logic worked across the board rather than patch this in at the transport?

Reply:
Right now pyAMQP uses 2 tasks receive_task and callback_task to receive, whereas uamqp technically uses one. This difference is the reason why async receive is in the transport layer. 

We can work on fixing pyAMQP to also use one thread and test everything to align the consumer logic. As tested so far it seems to cause issues receiving. However, in the meantime while still leaving it as 2 for now we can switch pyAMQP to also use `next_msg_in_buffer` like uamqp does to take a step in this direction of the two receives aligning and update the consumer logic for next release.

Reply:
@annatisch we definitely want to make pyamqp receive behave the same as uamqp ( Its on my semester list :) ). I took a shot at making it linear [here](https://github.com/Azure/azure-sdk-for-python/pull/31344/files) and as Libba mentioned there were issues around slowness with receives on a single partition and not working at all for multiple partitions. Figured it would be best to play it safe as making it linear would mean stress, perf & unit testing to make sure its on par with the current implementation.

Reply:
I'm not sure that "making pyamqp the same as uamqp" is the answer, they're different libraries with different designs that might work best when utilized differently. Though I prefer those differences to be kept inside their respective "transport" layers rather than it leaking out in the case of one and not the other.
If this is a problem that needs more thinking to solve, then by all means let's get a fix in place, as this bug is severe enough that I wouldn't want to wait another release cycle to fix - but let's make sure we document that by adding any inline-commenting as necessary.